### ISSUE TITLE
Inconsistent output from ActiveSupport::TimeZone.all

### ISSUE THREAD

**Issue by wonnage** [rails/rails#7245]
Due to caching, `ActiveSupport::TimeZone.all` returns different results if a non-ActiveSupport-supported zone was looked up first. 

```
ActiveSupport::TimeZone.all
# not in ActiveSupport::TimeZone::MAPPING, but still a valid zone
chicago = ActiveSupport::TimeZone['America/Chicago']
ActiveSupport::TimeZone.all.include?(chicago)
=> false
```

```
chicago = ActiveSupport::TimeZone['America/Chicago']
ActiveSupport::TimeZone.all.include?(chicago)
=> true
```

This affects `time_zone_options_for_select`, in that the `selected` arg of that function is a string matched to the names of zones in `ActiveSupport::TimeZone.all`. If your app stores timezones in TZInfo format, the helper may not generate an option tag for a recognized zone.

I see two ways around this:
1. Change the helper to recognize TZInfo identifiers
2. Update the zones cache when lazy-loading time zones.

Changing the helper might unintentionally change your data ("America/Chicago" would get converted to "Central Time"). It seems like Rails is opinionated about what zones it wants to use, so that might not be a big deal. The second approach avoids that problem, but still requires you to look up the alternate zone before it shows up in the list.

Thoughts?

**Comment by NilAndNull** [comment]
Hi,

This issue still lives and I don't know any workaround. :/

As standard time_zone_select has 146 timezones in it.
 
```
ActiveSupport::TimeZone.all.size
=> 146
```

If you happen to access one which isn't in the default list it gets added to the internal ActiveSupport::TimeZone.zones_map list
```
ActiveSupport::TimeZone["Europe/London"]
=> #<ActiveSupport::TimeZone:0x007faf3c5de1a0 @name="Europe/London", @utc_offset=nil, @tzinfo=#<TZInfo::TimezoneProxy: Europe/London>, @current_period=#<TZInfo::TimezonePeriod: #<TZInfo::TimezoneTransitionDefinition: #<TZInfo::TimeOrDateTime: 1414285200>,##<TZInfo::TimezoneOffset: 0,0,GMT>>,#<TZInfo::TimezoneTransitionDefinition: #<TZInfo::TimeOrDateTime: 1427590800>,#<TZInfo::TimezoneOffset: 0,3600,BST>>>>
```
 
However, zones_map is cached so we don't see any side effects as we'd already called #all previously

```
ActiveSupport::TimeZone.all.size
=> 146
``` 

But if we clear the caches (to pretend we hadn't accessed #all)
```
ActiveSupport::TimeZone.instance_variable_set("@zones", nil)
ActiveSupport::TimeZone.instance_variable_set("@zones_map", nil)
 ```

Then it appears
```
ActiveSupport::TimeZone.all.size 
=> 147
``` 

So now time_zone_select will include that extra option
 
Basically if you have any code which touches ActiveSupport::TimeZone#[]
And that happens *before* anything calls ActiveSupport::TimeZone.all
Then it will appear in the output of time_zone_select
Why this happen, I don't know. 

This is annoying if you auto-detect timezone from browser time, which come in TZInfo format (e.g. Europe/London).

If ActiveSupport::TimeZone.find_tzinfo method worked without creating this side-effect, that would solve the problem.

**Comment by jeremy** [comment]
`AS::TimeZone.all` changing depending on what happens to have been looked up is even stranger, asking for brittle, hard to reproduce bugs.

How about ensuring that it always faithfully returns the collection of time zones in MAPPING?

```ruby
def all
  @all ||= MAPPING.transform_values { |tz_identifier| find_tzinfo tz_identifier }
end
```

**Comment by y-yagi** [comment]
Fixed by #31176.
